File: lifehub-web/src/base/api/sync.ts
import apiClient from './client'
import type { ApiResponse } from './types'

export interface SyncChange {
  id: string
  entity_type: string
  entity_id: string
  action: string
  version: number
  synced_at: string
}

export async function getChanges(since: string, limit: number = 100): Promise<ApiResponse<SyncChange[]>> {
  const response = await apiClient.get('/sync/changes', { params: { since, limit } })
  return response.data
}

File: lifehub-backend/app/base/api/sync.py
"""Sync API routes."""

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.base.db.session import get_db
from app.base.services.sync_service import sync_service
from app.common.deps import get_current_user
from app.common.response import success_response

router = APIRouter(prefix="/api/sync", tags=["Sync"])


@router.get("/changes")
async def get_changes(
    since: str = Query(..., description="ISO timestamp to fetch changes since"),
    limit: int = Query(100, ge=1, le=1000),
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
):
    """Get all data changes since a given timestamp."""
    logs = await sync_service.get_changes_since(db, current_user.id, since, limit)
    return success_response(
        data=[
            {
                "id": str(log.id),
                "entity_type": log.entity_type,
                "entity_id": str(log.entity_id),
                "action": log.action,
                "version": log.version,
                "synced_at": log.created_at.isoformat(),
            }
            for log in logs
        ]
    )
